lib/puppet/provider/zfs_share/solaris.rb

```diff
@@ -0,0 +1,223 @@
+      allow_ip_read  = @resource[:allow_ip_read]
+      allow_ip_write = @resource[:allow_ip_write]
+      anon_user_id   = @resource[:anon_user_id]
+      nosub          = @resource[:nosub]
+      types          = ['none', 'sys', 'default']
+      if security.kind_of?(String)
+        security = [security]
+      end
+      types.reject!{ |x| security.include? x}
+      types.each do |t|
+        if share[/sec=#{t}/]
+          IO.popen(['zfs', 'set', '-c', "share=#{share_split[0, 3].join(',')},sec=#{t}", name])
+        end
+      end
+      sec = share[/sec=\w*(?=($|,))/]
+      if not allow_ip_read
+        if share =~ /,ro=[@|\*]/
+          IO.popen(['zfs', 'set', '-c', "share=#{share_split[0, 3].join(',')},#{sec},ro", name])
+        end
+      end
+      if not allow_ip_write
+        if share =~ /,rw=[@|\*]/
+          IO.popen(['zfs', 'set', '-c', "share=#{share_split[0, 3].join(',')},#{sec},rw", name])
+        end
+      end
+      if not nosub
+        if share =~ /nosub=true/
+          IO.popen(['zfs', 'set', '-c', "share=#{share_split[0, 3].join(',')},nosub", name])
+        end
+      end
+      if not anon_user_id
+        if share =~ /anon=\d+/
+          IO.popen(['zfs', 'set', '-c', "share=#{share_split[0, 3].join(',')},anon", name])
+        end
+      end
+      if protocol
+        if not protocol.include?('smb')
+          if share =~ /prot=smb/
+            IO.popen(['zfs', 'set', '-c', "share=#{share_split[0, 2].join(',')}prot=smb", name])
+          end
+        end
+      end
+    end
+  end
+
+  def purge_share
+    # "When the share to be set and the share on disk differ from one another first remove the share currently configured in its entirety.
+    # This allows changing the share title without manual intervention after initially setting it. The caveat of this parameter is that
+    # there's a potential for the NFS server connection on the remote end to be lost during the brief period that access is removed."
+    get_share = ['zfs', 'get', 'share', share_path()]
+    share = return_cmd(get_share)
+    if share.length > 0
+      share_split = share[-1].split()[2]
+      IO.popen(['zfs', 'set', '-c', "share=#{share_split.split(',')[0, 2].join(',')}", share_path()])
+    end
+  end
+
+  def exists?
+    # "Check is the share exists."
+    begin
+      name = share_path()
+      if File.directory?("/#{name}")
+        get_share_nfs = ['zfs', 'get', 'share.nfs', name]
+        get_share = ['zfs', 'get', 'share', name]
+        output = return_cmd(get_share_nfs)
+        share = return_cmd(get_share)
+        if output.length > 0
+          if output[-1].split()[2] == 'on'
+            if share.length > 0
+              check_for_stale_data()
+              share_ = share_cmd()
+              if share[-1].split()[2] == share_
+                true
+              end
+            end
+          end
+        end
+      end
+    end
+    rescue Puppet::ExecutionFailure
+      false
+  end
+
+  def create
```

Comment:
@mattkirby practical experience...

```
root@stor-filer5-test:~# zfs get share tank01/foo1
NAME         PROPERTY  VALUE  SOURCE
tank01/foo1  share     name=tank01_foo1,path=/tank01/foo1,prot=nfs,log=global,sec=none,rw=@10.28.92.0/24:@10.28.96.0/24  local
root@stor-filer5-test:~# zfs get share.nfs.sec.none.rw tank01/foo1
NAME         PROPERTY               VALUE                          SOURCE
tank01/foo1  share.nfs.sec.none.rw  @10.28.92.0/24:@10.28.96.0/24  local
```
vs.
```
root@stor-backups1:~# zfs get share backup-tank/backup/wasp.delivery.puppetlabs.net_osmirror
NAME                                                      PROPERTY  VALUE  SOURCE
backup-tank/backup/wasp.delivery.puppetlabs.net_osmirror  share     name=backup-tank_backup_wasp.delivery.puppetlabs.net_osmirror,path=/backup-tank/backup/wasp.delivery.puppetlabs.net_osmirror,prot=nfs,anon=50555,nosub=true,sec=none,rw=@10.32.22.32/32  local
root@stor-backups1:~# zfs get share.nfs.sec.none.rw backup-tank/backup/wasp.delivery.puppetlabs.net_osmirror
NAME                                                      PROPERTY               VALUE  SOURCE
backup-tank/backup/wasp.delivery.puppetlabs.net_osmirror  share.nfs.sec.none.rw         default
```
You'll just notice that explicit lookup on stor-backups1 is returning nil.